Parallel Tuning Strategies - MPI Communication for all existing tuning strategies

wothie opened the pull request:
# Description

Provides a new TuningStrategy (MPIParallelizedStrategy) which wraps around another TuningStrategy and handles its communication to other ranks. More specifically it splits the provided options among all ranks, lets them tune locally according to their strategy and then selects the optimal configuration across all ranks.
For this purpose it includes a new compiler option, new Options (MPIStrategyOption) and a wrapper for basic mpi calls to reduce the number of ifdefs. It also provides the AutoPasConfigurationCommunicator.h to handle distributing and comparing configurations across ranks, which is used in MPIParallelizedStrategy.
Due to how communication is done, it is pretty much limited to a homogeneous domain.

## Todo
- [x] Adapt ContributionGuidelines if needed
- [X] Try implementing Active Harmony without MPI. Result: Doesn't work as client count needs to be set during session start.

## Related Pull Requests

- Basically the completed version of #446 


## Resolved Issues

- [X] #410 mostly (except for the heterogeneous case)

# How Has This Been Tested?

- [X] locally with newly written tests and mostly sph-main-mpi
- [x] with all examples
- [x] on a cluster
- [x] very thoroughly

SteffenSeckler commented:
@wothie you have somehow broken the tests `Generated/TuningStrategyTest.testSearchSpaceOneOption/*`

SteffenSeckler commented on tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:
```diff
@@ -0,0 +1,153 @@
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  finiteCellSizeFactorsSetup(mpiParallelizedStrategy);
+
+  EXPECT_EQ(mpiParallelizedStrategy.getCurrentConfiguration(),
+            Configuration(ContainerOption::linkedCells, 1.0, TraversalOption::lc_sliced, LoadEstimatorOption::none,
+                          DataLayoutOption::aos, Newton3Option::enabled));
```
why? describe as comment

SteffenSeckler commented on tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:
```diff
@@ -0,0 +1,153 @@
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  finiteCellSizeFactorsSetup(mpiParallelizedStrategy);
+
+  EXPECT_LE(mpiParallelizedStrategy.getCurrentConfiguration().cellSizeFactor, 1.1);
```
explain, why.

SteffenSeckler commented on tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:
```diff
@@ -0,0 +1,153 @@
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  auto smallestLocalCellSizeFactor = infiniteCellSizeFactorSetup(mpiParallelizedStrategy);
+
+  EXPECT_LE(mpiParallelizedStrategy.getCurrentConfiguration().cellSizeFactor, smallestLocalCellSizeFactor);
```
explain why

SteffenSeckler commented on tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:
```diff
@@ -0,0 +1,153 @@
+                                                         oneTraversal, oneLoadEstimator, oneDataLayout, oneNewton3);
+  finiteCellSizeFactorsSetup(mpiParallelizedStrategy);
+
+  EXPECT_EQ(mpiParallelizedStrategy.getCurrentConfiguration(),
```
explain why

SteffenSeckler commented on tests/MPIParallelAutoPasTests/tests/selectors/tuningStrategy/MPIParallelizedStrategyTest.cpp:
```diff
@@ -0,0 +1,153 @@
+
+  // BayesianSearch seems to not be absolutely optimal in this case.
+  double error = 0.01;
+  EXPECT_LE(mpiParallelizedStrategy.getCurrentConfiguration().cellSizeFactor, smallestCellSizeFactor + error);
```
why?

SteffenSeckler commented on tests/MPIParallelAutoPasTests/tests/utils/AutoPasConfigurationCommunicatorTest.cpp:
```diff
@@ -0,0 +1,219 @@
+/**
```
shortly explain each test:
what are you testing and why do you expect, what you checking.

SteffenSeckler reviewed (changes requested):
see comments.

SteffenSeckler commented on src/autopas/selectors/tuningStrategy/ActiveHarmony.h:
```diff
@@ -6,12 +6,21 @@
 
 #pragma once
 
+#include <unistd.h>
```
you can find `getenv` in `cstdlib` and `snprintf` in `cstdio`

Njan0 commented on src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h:
```diff
@@ -166,9 +167,14 @@ class BayesianClusterSearch : public TuningStrategyInterface {
+  inline long getEvidence(Configuration configuration) const override { return _traversalTimes.at(configuration); }
+
   inline void reset(size_t iteration) override {
+    _gaussianCluster.clear();
```
This is intended. GaussianCluster needs a bunch of evidence to work properly. We keep the evidence over several tuning phases to avoid constant resampling.
```suggestion
```

SteffenSeckler replied:
when do you actually clear this?

Njan0 replied:
Currently we only decay their influence over time. Theoretically we could delete them if their influence becomes negligible. But we have not yet encountered any problems in our tests yet. To cause performance problems you probably have to set maxEvidence very high.

However, if this is a problem, the best option would probably be to restart the whole strategy after a few tuning phases.

```suggestion
    if (++phasesSinceLastReset >= phasesPerReset) {
        _gaussianCluster.clear();
        _firstTuningPhase = true;
        phasesSinceLastReset = 0;
    }
```

Njan0 replied:
On that note you probably need to add the following after resetting fullSearch.

`_currentConfig = _fullSearch.getCurrentConfiguration();`

https://github.com/AutoPas/AutoPas/blob/b2f476522ed55acd098edffd951066d581e918c1/src/autopas/selectors/tuningStrategy/BayesianClusterSearch.h#L188-L190

wothie replied:
@Njan0 do I understand it correctly that the last two suggestions are not really necessary?
I just assumed that the gaussianCluster should also be reset, but if you say there are no problems by not calling clear() than I don't see any either.
Is your last comment also related to this or is it a general suggestion?

Njan0 replied:
My last comment would generally be the correct implementation. But this is only a problem if someone calls reset() although tune() still returns true. So something that currently should never happen. But maybe that will change one day. ;)

wothie commented on src/autopas/selectors/tuningStrategy/ActiveHarmony.h:
```diff
@@ -29,30 +38,33 @@ class ActiveHarmony : public TuningStrategyInterface {
         _allowedDataLayoutOptions(allowedDataLayoutOptions),
         _allowedNewton3Options(allowedNewton3Options),
-        _currentConfig() {
-    AutoPasLog(debug, "Reducing traversal and container options to possible combinations");
-    for (auto &traversalOption : _allowedTraversalOptions) {
-      auto compatibleContainers = compatibleTraversals::allCompatibleContainers(traversalOption);
-      if (compatibleContainers.empty() or
-          allowedContainerOptions.find(*compatibleContainers.begin()) == allowedContainerOptions.end()) {
-        _allowedTraversalOptions.erase(traversalOption);
-      } else {
-        _allowedContainerOptions.emplace(*compatibleContainers.begin());
-      }
-    }
```
Got replaced by a call to utils::AutoPasConfigurationCommunicator::distributeConfigurations(...), which essentially does the same thing.

wothie commented on src/autopas/selectors/tuningStrategy/RandomSearch.h:
```diff
@@ -46,23 +46,49 @@ class RandomSearch : public TuningStrategyInterface {
-  inline void addEvidence(long time, size_t iteration) override { _traversalTimes[_currentConfig] = time; }
+  inline void addEvidence(long time, size_t iteration) override {
+    auto testingConfig = Configuration(_currentConfig);
+    testingConfig.cellSizeFactor = 1;
+    if (_traversalTimes.find(testingConfig) == _traversalTimes.end()) {
+      ++_numTestedConfigsNoCSF;
+    }
+    _traversalTimes[_currentConfig] = time;
```
Came from a conversation I had with @FG-TUM 
Essentially I couldn't find a reliable way to figure out how many valid configs there could still exist for infinite CSFs, so I ignore them alltogether.
This is not perfect and we might want to have another talk about a cleaner solution at some point.

SteffenSeckler replied:
can you document this? -> write a comment.

SteffenSeckler commented on src/autopas/selectors/tuningStrategy/MPIParallelizedStrategy.h:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  [[nodiscard]] const Configuration &getCurrentConfiguration() const override {
+    // cellSizeFactor == -1 iff a config is invalid
+    if (_optimalConfiguration.cellSizeFactor == -1) {
```
Maybe you could add the method `isValid()` to the Configuration class. Might help to make this clearer.

wothie replied:
Do you think ```hasValidValues()``` would get the point across properly?

SteffenSeckler replied:
seems fine

wothie commented on src/autopas/selectors/tuningStrategy/ActiveHarmony.h:
```diff
@@ -29,30 +38,33 @@ class ActiveHarmony : public TuningStrategyInterface {
+   * @param mpiStrategyOption
+   * @param comm Default value shouldn't be used. Only provided to not have to set it as first parameter.
    */
   ActiveHarmony(const std::set<ContainerOption> &allowedContainerOptions = ContainerOption::getAllOptions(),
```
The tests also call the constructor. Since I generally use ```mpi.h``` in the tests and ```MPIWrapper.h``` inside of AutoPas, I need the default values to be able to call the constructor at all.

SteffenSeckler replied:
Just tested this:
removing ALL defaults for this constructor and adding MPI_COMM_WORLD (or AUTOPAS_MPI_COMM_WORLD) as last argument for the constructor (make_unique) in `MPIParallelizedStrategyTest.cpp` works.

SteffenSeckler commented:
fyi: if you merge the same remote branch into your own branch, it is often recommended to use rebase instead of merge.
This will apply the remote changes first and replay your local changes/commits afterward.

wothie commented:
@SteffenSeckler the review suggestions have all been addressed so you might look over the non-trivial things again, if you want to.
(is there a "request review of changes since last review"-button somewhere?)

SteffenSeckler commented:
@wothie there's a re-request review at the top ;)

SteffenSeckler reviewed (approved):
only a few minor things.
Please adapt them first. Then we can merge the PR.

wothie commented on src/autopas/selectors/tuningStrategy/ActiveHarmony.h:
```diff
@@ -361,7 +364,7 @@ void ActiveHarmony::resetHarmony() {
     }
 
-    if (!_nonLocalServer || rank == 0) {
+    if (not _nonLocalServer || rank == 0) {
```
Should I also change other uses of ```||``` and ```&&``` to ```or``` and ```and```?

SteffenSeckler replied:
if you find any, please do :)
